src/utilities/hash.ts

```diff
@@ -0,0 +1,59 @@
+//https://stackoverflow.com/questions/40958727/javascript-generate-unique-number-based-on-string
+// https://github.com/bryc/code/blob/master/jshash/experimental/cyrb53.js
+
+import type { ExternalEventDB, ExternalEventPkey } from '../types/external-event';
+import type { ExternalSourcePkey, ExternalSourceSlim } from '../types/external-source';
+
+/*
+    cyrb53a (c) 2023 bryc (github.com/bryc)
+    License: Public domain. Attribution appreciated.
+    The original cyrb53 has a slight mixing bias in the low bits of h1.
+    This shouldn't be a huge problem, but I want to try to improve it.
+    This new version should have improved avalanche behavior, but
+    it is not quite final, I may still find improvements.
+    So don't expect it to always produce the same output.
+*/
+export const cyrb53a = function (str: string, seed = 0) {
+  let h1 = 0xdeadbeef ^ seed,
+    h2 = 0x41c6ce57 ^ seed;
+  for (let i = 0, ch; i < str.length; i++) {
+    ch = str.charCodeAt(i);
+    h1 = Math.imul(h1 ^ ch, 0x85ebca77);
+    h2 = Math.imul(h2 ^ ch, 0xc2b2ae3d);
+  }
+  h1 ^= Math.imul(h1 ^ (h2 >>> 15), 0x735a2d97);
+  h2 ^= Math.imul(h2 ^ (h1 >>> 15), 0xcaf649a9);
+  h1 ^= h2 >>> 16;
+  h2 ^= h1 >>> 16;
+  return 2097152 * (h2 >>> 0) + (h1 >>> 11);
+};
+
+export function getRowIdExternalSourceSlim(externalSourceSlim: ExternalSourceSlim) {
+  // https://stackoverflow.com/questions/40958727/javascript-generate-unique-number-based-on-string
+  return cyrb53a(externalSourceSlim.pkey.derivation_group_name + externalSourceSlim.pkey.key);
```

Comment:
Why a hash vs just the actual string?

Reply:
I believe it was because tables needed an actual integer id for indexing purposes, instead of a string!

Reply:
I think it should be able to take a `RowId` type, which is a string or number. If the string wasn't working, then that's maybe a bug that should be looked at in the future.

Reply:
I think the place where problems start cropping up is with `timeline.ts` (in `types`). We have `selectedActivityDirectiveId`, `selectedExternalEventId`, and `selectedSpanId`, all of which are `number | undefined`, and this is a pretty widely used event type definition which expects that. We have a few functions dispatching this event, and to get an id for a row, they need a number. We _could_ update those types to be of type `RowId`, but that might cascade a lot of changes, including:

- `TimelinePanel.svelte`, anywhere where `selectedExternalEventId` is found,
- `LayerDiscrete.svelte`, again where `selectedExternalEventId` is found (it seems that we could change selectedExternalEventId to accept a `RowId`, but then, as mentioned earlier, it would probably be necessary for uniformity to update `selectedActivityDirectiveId` and `selectedSpanId`)
- 'Row.svelte`, in how we have implemented `idToColorMaps`, which expects a numeric id as of now to retrieve colors

At the moment, it seems like what we have is the simplest way of doing it but if a refactor seems necessary, we could do that.

Reply:
Sorry. I didn't mean that the selected IDs for activities and spans needed to be refactored to match. `RowId` is a table specific type, so i wouldn't expect it to be used outside of that context. I think the crux of the ids is that they're unique within their context. So, I don't think they all have to be numbers or strings as long as they don't collide.

I do think that the benefit to keeping it as a string is for ease of debugging. It doesn't look like there's a way to decode the ID into the parts that made it up, so that would just be an extra step in trying to figure out with source was selected while debugging.

Reply:
I think where I was going with this is the selectedIds and the row ids overlap, meaning we always favor numbers, as far as I could tell in the current implementation. I think I'd agree as far as debugging, but making this change would affect more than just tables.

Reply:
After a second search through it appears that we could phase out the numeric id functions surrounding external sources and make those return strings, but those for events, because they do play with selectedEventId, and get passed/dispatched to places where a numeric id is expected, likely cannot.

Reply:
> I think where I was going with this is the selectedIds and the row ids overlap, meaning we always favor numbers, as far as I could tell in the current implementation

Sorry, we unfortunately have two meanings for a "row" haha. When you say "row id", do you mean a table row or a timeline row?

Reply:
Per our call earlier, I have now updated the codebase to make use of an `ExternalEventId` type, which is presently explicitly typed as a string. However, should this change, it can very easily be propagated across the codebase.

For reference, the nature of the issue is as follows:
- for easy referencing and selection of events, activities, or spans in a timeline, some sort of id is typically needed for these types.
- for activities and spans, this id is a number type. The reason, which was not mentioned earlier but makes the solution very clear, is so that when comparisons must be made between the spans/activities in a row (_which come from the database_) against the selected id, the id being used for comparison lines up with the id in the database. This id, in the database, is a number.
- such is not true for events - they have a composite key in the database, which to be condensed to a single type would require string concatenation to produce a single id - a string.
- we had used a hash of this string to force it to a number to encourage uniformity with the number ids used by activities and spans, however it has now become clear that this entirely defeats the purpose; that we are using numbers is not the issue. What matters is our id type match what is in the database.
- the database id type for an external event is a composite key, which we mentioned earlier should be compiled into a single string for easy use/referencing. We have now introduced a type, `ExternalEventId`, that is defined as a string and replaced every instance of the id being a number (a hash of the aforementioned string) with just the string itself. This is what can be found in [this commit here](https://github.com/NASA-AMMOS/aerie-ui/pull/1396/commits/0f7a86c8dc9f0b8988042e0c9a3c20d15af89454).

Reply:
Note also; we must use a string for `ExternalEventId` instead of directly using the composite `ExternalEventPkey` type. This is because there are some places, such as `LayerDiscrete`, where we need to use this ID type in a record (i.e. `IdToColorMap`). Records admit keys of type `string | number | symbol`, so an `ExternalEventPkey` would not work. Thus we _can_ confirm that this concatenation step to a string is necessary.